packages/core/src/domain/report/reportObservable.ts

```diff
@@ -0,0 +1,118 @@
+import { toStackTraceString } from '../../tools/error'
+import { mergeObservables, Observable } from '../../tools/observable'
+import { DOM_EVENT, includes, addEventListener } from '../../tools/utils'
+import { monitor } from '../internalMonitoring'
+import type { Report, BrowserWindow, ReportType } from './browser.types'
+
+export const RawReportType = {
+  intervention: 'intervention',
+  deprecation: 'deprecation',
+  cspViolation: 'csp_violation',
+} as const
+
+export type RawReportType = typeof RawReportType[keyof typeof RawReportType]
+
+export interface RawReport {
+  type: RawReportType
+  subtype: string
+  message: string
+  stack?: string
+}
+
+export function initReportObservable(apis: RawReportType[]) {
+  const observables: Array<Observable<RawReport>> = []
+
+  if (includes(apis, RawReportType.cspViolation)) {
+    observables.push(createCspViolationReportObservable())
+  }
+
+  const reportTypes = apis.filter((api: RawReportType): api is ReportType => api !== RawReportType.cspViolation)
+  if (reportTypes.length) {
+    observables.push(createReportObservable(reportTypes))
+  }
+
+  return mergeObservables<RawReport>(...observables)
+}
+
+function createReportObservable(reportTypes: ReportType[]) {
+  const observable = new Observable<RawReport>(() => {
+    if (!(window as BrowserWindow).ReportingObserver) {
+      return
+    }
+
+    const handleReports = monitor((reports: Report[]) =>
+      reports.forEach((report) => {
+        observable.notify(buildRawReportFromReport(report))
+      })
+    )
+
+    const observer = new (window as BrowserWindow).ReportingObserver!(handleReports, {
+      types: reportTypes,
+      buffered: true,
+    })
+
+    observer.observe()
+    return () => {
+      observer.disconnect()
+    }
+  })
+
+  return observable
+}
+
+function createCspViolationReportObservable() {
+  const observable = new Observable<RawReport>(() => {
+    const handleCspViolation = monitor((event: SecurityPolicyViolationEvent) => {
+      observable.notify(buildRawReportFromCspViolation(event))
+    })
+
+    const { stop } = addEventListener(document, DOM_EVENT.SECURITY_POLICY_VIOLATION, handleCspViolation)
+
+    return stop
+  })
+  return observable
+}
+
+function buildRawReportFromReport({ type, body }: Report): RawReport {
+  return {
+    type,
+    subtype: body.id,
+    message: `${type}: ${body.message}`,
+    stack: buildStack(body.id, body.message, body.sourceFile, body.lineNumber, body.columnNumber),
+  }
+}
+
+function buildRawReportFromCspViolation(event: SecurityPolicyViolationEvent): RawReport {
+  const type = RawReportType.cspViolation
+  const message = `'${event.blockedURI}' blocked by '${event.effectiveDirective}' directive`
```

Comment:
After some thought, I think it is valuable to add the `originalPolicy` to the crafted message. Could be done like so:
`${event.blockedURI}' blocked by '${event.effectiveDirective}' directive of the policy "${event.originalPolicy}"`.

Reply:
@bcaudan @BenoitZugmeyer

Reply:
Some directives could list a large number of domains 🤔
I'd say that it could be added in the stack to keep a minimal message.
We should maybe also truncate the policy to avoid sending too much data.

Reply:
I added it to the stack. Let me know what you think

Reply:
LGTM!